transport/src/test/java/io/netty/channel/embedded/EmbeddedChannelTest.java

```diff
@@ -605,6 +608,91 @@ public void channelInactive(ChannelHandlerContext ctx) throws Exception {
         assertTrue(inactive.get());
     }
 
+    @Test
+    public void testCloseAutoRead() {
+        EmbeddedChannel channel = new EmbeddedChannel();
+        channel.config().setAutoRead(false);
+
+        Queue<Object> inboundMessages = channel.inboundMessages();
+
+        // write three element to this channel.
+        channel.writeInbound(1L, 2L, 3L);
+        assertEquals(0, inboundMessages.size());
+
+        // call `channel.read()` to fire `channelRead` event.
+        channel.read();
+        assertEquals(3, inboundMessages.size());
+        inboundMessages.clear();
+
+        // test `writeOneInbound`
+        channel.writeOneInbound(1L);
+        channel.flushInbound();
+
+        // since `autoRead` was still false, this will not fire `channel.read()`.
+        assertEquals(0, inboundMessages.size());
+
+        assertFalse(channel.finish());
+    }
+
+    @Test
+    public void testChannelPromiseWithAutoReadClosed() {
+        final List<Long> inboundMsgs = new ArrayList<Long>();
+        EmbeddedChannel channel = new EmbeddedChannel(new ChannelInboundHandlerAdapter() {
+
+            @Override
+            public void channelRead(ChannelHandlerContext ctx, Object msg) throws Exception {
+                inboundMsgs.add((Long) msg);
+            }
+        });
+        channel.config().setAutoRead(false);
+
+        // no msg since close autoRead.
+        assertEquals(0, inboundMsgs.size());
+
+        // channelFuture will not be completed since autoRead was closed.
+        ChannelFuture channelFuture = channel.writeOneInbound(1L, channel.newPromise());
+        assertFalse(channelFuture.isDone());
+
+        // open channel autoRead.
+        channel.config().setAutoRead(true);
+        assertEquals(1, inboundMsgs.size());
+        assertTrue(channelFuture.isDone());
+        assertTrue(channelFuture.isSuccess());
+
+        assertFalse(channel.finish());
+    }
+
+    @Test
+    public void testCloseChannelWithUnhandledMsg() {
+        final List<Long> inboundMsgs = new ArrayList<Long>();
+        EmbeddedChannel channel = new EmbeddedChannel(new ChannelInboundHandlerAdapter() {
+
+            @Override
+            public void channelRead(ChannelHandlerContext ctx, Object msg) throws Exception {
+                inboundMsgs.add((Long) msg);
+            }
+        });
+        channel.config().setAutoRead(false);
+
+        ChannelFuture channelFuture = channel.writeOneInbound(1L, channel.newPromise());
+
+        // no msg since close autoRead.
+        assertEquals(0, inboundMsgs.size());
+
+        Queue<Object> inboundMessages = channel.inboundMessages();
+        assertEquals(inboundMessages.size(), 0);
+        assertFalse(channel.finish());
+
+        try {
+            channelFuture.get();
+            fail();
+        } catch (InterruptedException e) {
+            fail();
+        } catch (ExecutionException e) {
+            Matchers.instanceOf(ClosedChannelException.class);
```

Comment:
this must be `assertThat(e.getCause(), Matchers.instanceOf(ClosedChannelException.class));`